### DISCUSSION TITLE
Versioning/Partitioning Assets with IO Manager?

### DISCUSSION THREAD

**Discussion by kiranhda** [dagster-io/dagster#25375] [Q&A]
Hi all! I was hoping to get a bit of input from the community - my team has been working on our custom S3 IO manager for the past week, and we were wondering - What's the best approach for versioning asset materializations in Dagster? 
We want to:

- Maintain latest versions for quick access
- Preserve historical versions
- Allow flexible input selection for downstream assets
- Support dev/prod environments and monthly versions of assets
- Possibly store run IDs in DynamoDB or in a config for easy retrieval

Should we implement a custom IO manager, use partitioning, or is there another built-in solution? Are partitions suitable for this use case (an entire version of an asset), or would they add unnecessary complexity?

**Comment by garethbrickman** [comment]
There is an experimental feature for [Asset versioning and caching](https://docs.dagster.io/guides/dagster/asset-versioning-and-caching#asset-versioning-and-caching)
